src/per/uart.cpp

```diff
@@ -34,135 +67,371 @@ struct uart_handle
+    GPIO_InitStruct.Speed = GPIO_SPEED_FREQ_LOW;
+
+    int per_num = 2 * (int)(config_.periph);
+
+    if(config_.pin_config.tx.port != DSY_GPIOX)
+    {
+        //check tx against periph
+        if(checkPinMatch(&GPIO_InitStruct, config_.pin_config.tx, per_num)
+           == Result::ERR)
+        {
+            return Result::ERR;
+        }
+
+        //setup tx pin
+        port                = dsy_hal_map_get_port(&config_.pin_config.tx);
+        GPIO_InitStruct.Pin = dsy_hal_map_get_pin(&config_.pin_config.tx);
+        tx                  = GPIO_InitStruct.Pin;
+        HAL_GPIO_Init(port, &GPIO_InitStruct);
+    }
+
+    if(config_.pin_config.rx.port != DSY_GPIOX)
+    {
+        //check rx against periph
+        if(checkPinMatch(&GPIO_InitStruct, config_.pin_config.rx, per_num + 1)
+           == Result::ERR)
+        {
+            return Result::ERR;
+        }
+
+        //setup rx pin
+        port                = dsy_hal_map_get_port(&config_.pin_config.rx);
+        GPIO_InitStruct.Pin = dsy_hal_map_get_pin(&config_.pin_config.rx);
+        rx                  = GPIO_InitStruct.Pin;
+        HAL_GPIO_Init(port, &GPIO_InitStruct);
+    }
+
+    return Result::OK;
+}
+
+UartHandler::Result UartHandler::Impl::DeInitPins()
+{
+    uint16_t pin;
+    port = dsy_hal_map_get_port(&config_.pin_config.tx);
+    pin  = dsy_hal_map_get_pin(&config_.pin_config.tx);
+    HAL_GPIO_DeInit(port, pin);
+    port = dsy_hal_map_get_port(&config_.pin_config.rx);
+    pin  = dsy_hal_map_get_pin(&config_.pin_config.rx);
+    HAL_GPIO_DeInit(port, pin);
+
+    return Result::OK;
+}
+
 // Callbacks
-static void UARTRxComplete(void)
+static void UARTRxComplete(UartHandler::Impl* impl)
 {
     size_t len, cur_pos;
     //get current write pointer
-    cur_pos = (uhandle.rx_size
-               - ((DMA_Stream_TypeDef*)uhandle.huart1.hdmarx->Instance)->NDTR)
-              & (uhandle.rx_size - 1);
+    cur_pos = (impl->rx_size
+               - ((DMA_Stream_TypeDef*)impl->huart1.hdmarx->Instance)->NDTR)
+              & (impl->rx_size - 1);
     //calculate how far the DMA write pointer has moved
-    len = (cur_pos - uhandle.rx_last_pos + uhandle.rx_size) % uhandle.rx_size;
+    len = (cur_pos - impl->rx_last_pos + impl->rx_size) % impl->rx_size;
     //check message size
-    if(len <= uhandle.rx_size)
+    if(len <= impl->rx_size)
     {
-        uhandle.dma_fifo_rx->Advance(len);
-        uhandle.rx_last_pos = cur_pos;
+        impl->dma_fifo_rx->Advance(len);
+        impl->rx_last_pos = cur_pos;
 #ifdef UART_RX_DOUBLE_BUFFER
         // Copy to queue fifo we don't want to use primary fifo to avoid
         // changes to the buffer while its being processed
         uint8_t processbuf[256];
-        uhandle.dma_fifo_rx->ImmediateRead(processbuf, len);
-        uhandle.queue_rx.Overwrite(processbuf, len);
+        impl->dma_fifo_rx->ImmediateRead(processbuf, len);
+        impl->queue_rx.Overwrite(processbuf, len);
 #endif
```

Comment:
I don't see any reason for this not to be a class function for the `Impl` class. You're basically making it that in a c-style way anyway.

You would just have the pointer to the Impl class call it's own RxComplete function instead of having the function have a pointer to the object (see sai.cpp for example).